src/execution_tree/primitives/shuffle_operation.cpp

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2018 Parsa Amini
+// Copyright (c) 2018 Hartmut Kaiser
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <phylanx/config.hpp>
+#include <phylanx/execution_tree/primitives/shuffle_operation.hpp>
+#include <phylanx/util/matrix_iterators.hpp>
+
+#include <hpx/include/lcos.hpp>
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+#include <hpx/throw_exception.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <blaze/Math.h>
+
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives
+{
+    ///////////////////////////////////////////////////////////////////////////
+    primitive create_shuffle_operation(hpx::id_type const& locality,
+        std::vector<primitive_argument_type>&& operands,
+        std::string const& name, std::string const& codename)
+    {
+        static std::string type("shuffle_operation");
+        return create_primitive_component(
+            locality, type, std::move(operands), name, codename);
+    }
+
+    match_pattern_type const shuffle_operation::match_data =
+    {
+        hpx::util::make_tuple("shuffle_operation",
+            std::vector<std::string>{"shuffle_operation(__1)", "'(__1)"},
+            &create_shuffle_operation, &create_primitive<shuffle_operation>)
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    shuffle_operation::shuffle_operation(
+            std::vector<primitive_argument_type>&& operands,
+            std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    ///////////////////////////////////////////////////////////////////////////
+    primitive_argument_type shuffle_operation::shuffle_1d(args_type && args) const
+    {
+        auto x = args[0].vector();
+        std::shuffle(x.begin(), x.end(), std::mt19937{std::random_device{}()});
+
+        return primitive_argument_type(std::move(x));
+    }
+
+    primitive_argument_type shuffle_operation::shuffle_2d(args_type&& args) const
+    {
+        auto x = args[0].matrix();
+        auto x_begin = util::matrix_row_iterator<decltype(x)>(x);
+        auto x_end = util::matrix_row_iterator<decltype(x)>(x, x.rows());
+        std::shuffle(x_begin, x_end, std::mt19937{ std::random_device{}() });
```

Comment:
Would you see a way not to create a new generator and not calling `std::random_device` for each invocation of this primitive?

Reply:
Done.